### PR TITLE
Otp

### PR DESCRIPTION
**Pull request by pratik-mazumdar-wemotive**
Added OTP Flow.

### CODE REVIEW

File: lsp/api/models/order.model.ts [hunk 1]
```diff
@@ -10,10 +11,11 @@ export interface IOrder extends Document {
   orderValue: number;
   lspOrderValue: number;
   txnId: string;
+  deliveryOtp: number;
```
Comment (wemakshaychavan) [inline lsp/api/models/order.model.ts:14]: Alphanumeric

File: lsp/api/models/order.model.ts [hunk 2]
```diff
@@ -10,10 +11,11 @@ export interface IOrder extends Document {
...
+  pickupOtp: number;
```
Comment (wemakshaychavan) [inline lsp/api/models/order.model.ts:15]: Alphanumeric

File: lsp/api/models/order.model.ts [hunk 3]
```diff
@@ -30,13 +32,27 @@ const OrderSchema: Schema<IOrder> = new Schema<IOrder>({
...
   },
   fulfillmentId: {
     type: String,
     ref: 'Fulfillment.FulfillmentId',
     required: true,
   },
+
+  pickupOtp: {
+    required: false,
+    type: Number,
```
Comment (wemakshaychavan) [inline lsp/api/models/order.model.ts:44]: Alphanumeric

File: lsp/api/models/order.model.ts [hunk 4]
```diff
@@ -30,13 +32,27 @@ const OrderSchema: Schema<IOrder> = new Schema<IOrder>({
...
+    max: 999999,
+    min: 100000,
+  },
+
+  deliveryOtp: {
+    required: false,
```
Comment (wemakshaychavan) [inline lsp/api/models/order.model.ts:50]: Alphanumeric

File: lsp/api/v1/services/ondc.service.ts [hunk 5]
```diff
@@ -802,9 +828,11 @@ class OndcService {
     await updateLog.save();
 
     let headers = {};
+    console.log('Sending update request to procol:', JSON.stringify(updateRequest));
     // Send the cancel request to the eCommerce protocol layer
     let httpRequest = new HttpRequest(getEnv('PROTOCOL_BASE_URL'), `/update`, 'post', updateRequest, headers);
-    await httpRequest.send();
+    const response = await httpRequest.send();
+    if (response.data.message.ack.status === Ack.NACK) throw new HttpError(response.status, response.data.error);
```
Comment (wemakshaychavan) [inline lsp/api/v1/services/ondc.service.ts:835]: First check for http code = 200

File: lsp/interfaces/index.ts [hunk 6]
```diff
@@ -74,6 +74,13 @@ export class ReadyToShipPayload {
   @IsOptional()
   @IsEnum(['yes', 'no'])
   readyToShip!: 'yes' | 'no';
+  @IsNumber()
+  @IsOptional()
+  pickupOtp!: number;
```
Comment (wemakshaychavan) [inline lsp/interfaces/index.ts:79]: Alphanumeric

File: lsp/interfaces/index.ts [hunk 7]
```diff
@@ -74,6 +74,13 @@ export class ReadyToShipPayload {
...
+
+  @IsNumber()
+  @IsOptional()
+  deliveryOtp!: number;
```
Comment (wemakshaychavan) [inline lsp/interfaces/index.ts:83]: Alphanumeric

File: lsp/utils/payload-builder.ts [hunk 8]
```diff
@@ -444,19 +446,35 @@ const readyToShip = (onConfirm: OnConfirm, logisticsMessageId: UUID) => {
...
+                  long_desc: 'Please contact respective person and confirm before taking the parcel',
+                },
+                authorization: {
+                  type: 'OTP',
+                  token: orders.pickupOtp.toString(),
+                  valid_from: timeStamp.toISOString(),
+                  valid_to: new Date(timeStamp.setDate(timeStamp.getDate() + 1)).toISOString(),
+                },
+              }
+            : undefined,
```
Comment (wemakshaychavan) [inline lsp/utils/payload-builder.ts:463]: default instructions missing

File: lsp/utils/payload-builder.ts [hunk 9]
```diff
@@ -444,19 +446,35 @@ const readyToShip = (onConfirm: OnConfirm, logisticsMessageId: UUID) => {
...
+                  short_desc: orders.deliveryOtp.toString(),
+                },
+                authorization: {
+                  type: 'OTP',
+                  token: orders.deliveryOtp.toString(),
+                  valid_from: timeStamp.toISOString(),
+                  valid_to: new Date(timeStamp.setDate(timeStamp.getDate() + 1)).toISOString(),
+                },
+              }
+            : undefined,
```
Comment (wemakshaychavan) [inline lsp/utils/payload-builder.ts:477]: Default instructons missing

File: lsp/api/v1/services/ondc.service.ts [hunk 10]
```diff
@@ -42,9 +43,10 @@ import Payment from '../../models/payment';
 import { DuplicateRecordFoundError } from '../../errors';
 import mapCity from '../../../utils/mapCityCode';
 import redisService from './redis.service';
+import { clearInterval } from 'node:timers';
 
 class OndcService {
-  async quote(payload: Payload) {
+  async quote(payload: any) {
```
Comment (wemakshaychavan) [inline lsp/api/v1/services/ondc.service.ts:49]: why ANY?

File: lsp/api/v1/services/ondc.service.ts [hunk 11]
```diff
@@ -90,15 +93,28 @@ class OndcService {
...
+      });
+
+      filteredItems.forEach((filteredItem) => {
+        immediateDeliveryItems.push({ ...filteredItem, bpp_id: item.context.bpp_id });
+      });
     }
-    const bppIds = onSearchs.map((onSearchs: any) => JSON.parse(onSearchs).context.bpp_id);
-    console.log('All bpp_ids:====>', bppIds);
+
+    console.log('All bpp_ids:====>', immediateDeliveryItems);
```
Comment (wemakshaychavan) [inline lsp/api/v1/services/ondc.service.ts:112]: Variable is wrong - we will be doing all type of delivery

File: lsp/api/v1/services/ondc.service.ts [hunk 12]
```diff
@@ -528,224 +547,228 @@ class OndcService {
...
+                    const response = await axios.post(
+                      `${getEnv('RETAIL_CLIENT_WEBHOOK_ENDPOINT')}/order/${payload.logisticsOrderId}/ship`,
+                    );
+                    if (response.status === 200) console.log('Ship request sent successfully');
+                    else console.log('Ship request failed');
                   }
+                  return resolve(lspOnConfirmResponse);
                 } catch (error) {
-                  reject(error);
+                  return reject(error);
```
Comment (wemakshaychavan) [inline lsp/api/v1/services/ondc.service.ts:746]: Do we really need to return a reject?

File: lsp/interfaces/index.ts [hunk 13]
```diff
@@ -67,13 +69,17 @@ export class IssueCancelPayload {
...
   @IsOptional()
   @IsBoolean()
   onNetworkLogistics: Boolean = true;
+
+  @IsString()
   @IsOptional()
-  @IsEnum(['yes', 'no'])
-  readyToShip!: 'yes' | 'no';
+  @Length(6)
+  merchantCode!: string;
```
Comment (wemakshaychavan) [inline lsp/interfaces/index.ts:82]: We need both pickup and drop code.

Comment (pratik-mazumdar-wemotive) [reply lsp/interfaces/index.ts:82]: I have been taking those data from the create-order as you said yestarday. This is for non-otp flow.

File: lsp/api/v1/services/ondc.service.ts (hunk 12 above)
Comment (pratik-mazumdar) [reply lsp/api/v1/services/ondc.service.ts:746]: removed.

File: lsp/api/v1/services/ondc.service.ts (hunk 10 above)
Comment (pratik-mazumdar-wemotive) [reply lsp/api/v1/services/ondc.service.ts:49]: Fixed.

File: lsp/api/v1/services/ondc.service.ts (hunk 5 above)
Comment (pratik-mazumdar-wemotive) [reply lsp/api/v1/services/ondc.service.ts:824]: Fixed.

File: lsp/api/models/order.model.ts (hunk 4 above)
Comment (pratik-mazumdar-wemotive) [reply lsp/api/models/order.model.ts:50]: Fixed

**Pull request merged**